### PR TITLE
Initial admin panel with bulk create and manage monuments

### PR DESCRIPTION
**Pull request by benvogler**
Demo: https://gfycat.com/flusteredsmartbluet

### CODE REVIEW

File: client/src/components/AdminPanel/AdminPanel.js [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+                </div>
+                <div className="viewport">
+                    <ProtectedRoute exact path="/panel" component={AdminPanelHome} customProps={{role}}/>
+                    <ProtectedRoute exact path="/panel/bulk" component={MonumentBulkCreatePage}/>
+                    <ProtectedRoute exact path="/panel/manage/monuments" component={ManageMonumentsPage}/>
+                    <ProtectedRoute exact path="/panel/manage/monuments/search" component={ManageMonumentsPage} customProps={{mode: 'search'}}/>
+                    <ProtectedRoute exact path="/panel/manage/monuments/monument/:monumentId" component={ManageMonumentsPage} customProps={{mode: 'monument'}}/>
+                    <ProtectedRoute exact path="/panel/manage/monuments/monument/update/:monumentId" component={ManageMonumentsPage} customProps={{mode: 'update'}}/>
+                </div>
+                <div className="right"/>
```
Comment (njd9419) [inline client/src/components/AdminPanel/AdminPanel.js]: Any reason to have an empty `div` here?

File: client/src/components/AdminPanel/AdminPanelHome/AdminPanelHome.js [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+        return (
+            <div className="home">
+                <Card>
+                    <Card.Header>
+                        <Card.Title>
+                            Welcome to the {role} Panel
+                        </Card.Title>
+                    </Card.Header>
+                    <Card.Body>
+                        You can perform advanced actions such as bulk creating monuments or memorials from this page. Please see the sidebar to the left for advanced actions.
```
Comment (njd9419) [inline client/src/components/AdminPanel/AdminPanelHome/AdminPanelHome.js]: Does this read more clearly if it's `You can perform advanced actions such as bulk creating monuments and memorials from this page...`. (Swap "or" for "and").

File: client/src/components/AdminPanel/ManageMonuments/ManageMonument/ManageMonument.js [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+        return (
+            <div className="manage-monument">
+                <About monument={monument} contributions={monument.contributions} references={monument.references}
+                       header={monument.title} showHiddenFields hideExportToCSV hideTitle/>
+                {(monument.images && monument.images.length) ?
+                    <div className="images">
+                        {monument.images.map(image => (
+                            <div key={image.id} className="image" style={{backgroundImage: `url(${image.url})`}}/>
+                        ))}
+                    </div> : <></>
```
Comment (njd9419) [inline client/src/components/AdminPanel/ManageMonuments/ManageMonument/ManageMonument.js]: Couldn't the false case of this ternary be `</>`? Or can you not do that with an empty tag?

File: client/src/components/AdminPanel/ManageMonuments/ManageMonument/ManageMonument.js [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+        );
+    }
+
+    renderConfirmToggleActiveModal() {
+        const { toggleActiveModalOpen } = this.state;
+        const { monument } = this.props;
+
+        return (
+            <div onClick={e => e.stopPropagation()}>
+                <Modal show={toggleActiveModalOpen} onHide={() => this.setState({toggleActiveModalOpen: false})}>
```
Comment (njd9419) [inline client/src/components/AdminPanel/ManageMonuments/ManageMonument/ManageMonument.js]: The `onHide` here is clean. Wish I would have done that more instead if creating another function for it.

File: client/src/components/AdminPanel/Sidebar/Sidebar.js [hunk 5]
```diff
@@ -0,0 +1,32 @@
+import * as React from 'react';
+import './Sidebar.scss';
+import {Role} from '../../../utils/authentication-util';
```
Comment (njd9419) [inline client/src/components/AdminPanel/Sidebar/Sidebar.js]: Minor: Should be `{ Role }` (spacing)

File: client/src/components/AdminPanel/Sidebar/Sidebar.js [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+import {NavLink} from 'react-router-dom';
```
Comment (njd9419) [inline client/src/components/AdminPanel/Sidebar/Sidebar.js]: Minor: Should be `{ NavLink }` (spacing)

File: client/src/components/AdminPanel/Sidebar/Sidebar.js [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+
+export default class Sidebar extends React.Component {
+
+    get links() {
+        return [
+            {name: 'Home', icon: 'home', route: '/panel', exact: true},
+            {name: 'Bulk Create', icon: 'cloud_upload', route: '/panel/bulk'},
+            {name: 'Manage Monuments', icon: 'account_balance', route: '/panel/manage/monuments', roles: Role.RESEARCHER_OR_ABOVE},
+            {name: 'Manage Users', icon: 'person', route: '/panel/manage/users'}
```
Comment (njd9419) [inline client/src/components/AdminPanel/Sidebar/Sidebar.js]: Shouldn't this have `roles: Role.ADMIN` or something similar?

File: client/src/components/Monument/Gallery/Gallery.js [hunk 8]
```diff
@@ -168,7 +168,7 @@ export default class Gallery extends React.Component {
                         <div className="d-flex justify-content-center">
                             <img src={selectedImage.url} alt="large"/>
                         </div>
-                        <p className="caption">Image description...</p>
+                        {/*<p className="caption">Image description...</p>*/}
```
Comment (njd9419) [inline client/src/components/Monument/Gallery/Gallery.js]: Leftover code or something?

File: client/src/constants.js [hunk 9]
```diff
@@ -105,6 +105,12 @@ export const FETCH_MONUMENT_UPDATE_ERROR = 'FETCH_MONUMENT_UPDATE_ERROR';
 export const UPDATE_MONUMENT_PENDING = 'UPDATE_MONUMENT_PENDING';
 export const UPDATE_MONUMENT_SUCCESS = 'UPDATE_MONUMENT_SUCCESS';
 export const UPDATE_MONUMENT_ERROR = 'UPDATE_MONUMENT_ERROR';
+export const TOGGLE_MONUMENT_IS_ACTIVE_PENDING = 'TOGGLE_MONUMENT_IS_PENDING';
```
Comment (njd9419) [inline client/src/constants.js]: Should be `TOGGLE_MONUMENT_IS_ACTIVE_PENDING`.

File: client/src/constants.js [hunk 10]
```diff
@@ -105,6 +105,12 @@ export const FETCH_MONUMENT_UPDATE_ERROR = 'FETCH_MONUMENT_UPDATE_ERROR';
...
+export const TOGGLE_MONUMENT_IS_ACTIVE_SUCCESS = 'TOGGLE_MONUMENT_IS_SUCCESS';
```
Comment (njd9419) [inline client/src/constants.js]: Should be `TOGGLE_MONUMENT_IS_ACTIVE_SUCCESS`.

File: client/src/constants.js [hunk 11]
```diff
@@ -105,6 +105,12 @@ export const FETCH_MONUMENT_UPDATE_ERROR = 'FETCH_MONUMENT_UPDATE_ERROR';
...
+export const TOGGLE_MONUMENT_IS_ACTIVE_ERROR = 'TOGGLE_MONUMENT_IS_ERROR';
```
Comment (njd9419) [inline client/src/constants.js]: Should be `TOGGLE_MONUMENT_IS_ACTIVE_ERROR`.

File: client/src/containers/ProtectedRoute/ProtectedRoute.js [hunk 12]
```diff
@@ -14,26 +14,31 @@ class ProtectedRoute extends React.Component {
...
+                alert: message
             }
         }}/>);
 
         let loggedIn = session.user && session.user.role;
-        let render = (props) => (<Component {...props} />);
+        let render = (props) => (<Component {...props} {...customProps} />);
+
+        // If using exact path, match exactly, otherwise do some fuzzy checking
+        // TODO: Verify that this fuzzy checking is good enough
```
Comment (njd9419) [inline client/src/containers/ProtectedRoute/ProtectedRoute.js]: Is it good enough?

File: client/src/pages/AdminPage/ManageMonumentsPage/ManageMonumentsPage.js [hunk 13]
```diff
@@ -0,0 +1,60 @@
...
+    }
+
+    handleDeleteMonument() {
+        const { monument, dispatch } = this.props;
+        dispatch(deleteMonument(monument.id));
+    }
+
+    render() {
+        const { mode, monument, deleteMonument } = this.props;
+        console.log(deleteMonument);
```
Comment (njd9419) [inline client/src/pages/AdminPage/ManageMonumentsPage/ManageMonumentsPage.js]: Minor: Debug print.

File: client/src/components/AdminPanel/ManageMonuments/ManageMonument/ManageMonument.js (hunk 3 above)
Comment (benvogler) [reply client/src/components/AdminPanel/ManageMonuments/ManageMonument/ManageMonument.js]: Yea it doesn't work with the empty tag. I could probably just return null though

File: client/src/components/AdminPanel/Sidebar/Sidebar.js (hunk 6 above)
Comment (benvogler) [reply client/src/components/AdminPanel/Sidebar/Sidebar.js]: My IntelliJ settings got all messed up when I was changing them for 344-rest 😢

File: client/src/components/Monument/Gallery/Gallery.js (hunk 8 above)
Comment (benvogler) [reply client/src/components/Monument/Gallery/Gallery.js]: Yea, I noticed this was laying around and we never actually added captions to images yet so for now I'm hiding it

File: client/src/containers/ProtectedRoute/ProtectedRoute.js (hunk 12 above)
Comment (benvogler) [reply client/src/containers/ProtectedRoute/ProtectedRoute.js]: Seems to be... it basically allows `/monuments/:monumentId` for example to match to `/monuments/3`. I'm wondering if there's a good way to 100% verify that they match, maybe through react router...

File: client/src/components/Monument/Gallery/Gallery.js (hunk 8 above)
Comment (njd9419) [reply client/src/components/Monument/Gallery/Gallery.js]: Ah ok didn't realize that.

File: src/main/java/com/monumental/controllers/MonumentController.java [hunk 14]
```diff
@@ -51,17 +61,33 @@ public Monument createMonument(@RequestBody CreateMonumentRequest monumentReques
     }
 
     /**
-     * Get a Monument with the specified ID, if it exists
+     * Get a Monument with the specified ID, if it exists and is active or inactive depending on onlyActive
      * @param id - ID of the Monument to get
      * @param cascade - If true, loads all of the lazy-loaded collections associated with the Monument
+     * @param onlyActive - If true, a 404 will be returned if the specified Monument is inactive. If false, the monument
+     *                 will be returned regardless of whether it's active or inactive.
+     *                 If this is false then the user must be a partner or researcher to view
```
Comment (njd9419) [inline src/main/java/com/monumental/controllers/MonumentController.java]: Can you correct this comment to be `If this is false then the user must be a partner or above to view`?

File: src/main/java/com/monumental/controllers/MonumentController.java [hunk 15]
```diff
@@ -51,17 +61,33 @@ public Monument createMonument(@RequestBody CreateMonumentRequest monumentReques
...
      * @return Monument - The Monument with the specified ID, if it exists
-     * @throws ResourceNotFoundException - If a Monument with the specified ID does not exist
+     * @throws ResourceNotFoundException - If a Monument with the specified ID does not exist or onlyActive is true and isActive is false
+     * @throws AccessDeniedException - If trying to get an inactive monument without being a partner or researcher
```
Comment (njd9419) [inline src/main/java/com/monumental/controllers/MonumentController.java]: Same here: `If trying to get an inactive monument without being a partner or above`.

File: src/main/java/com/monumental/controllers/MonumentController.java [hunk 16]
```diff
@@ -72,12 +98,22 @@ public Monument getMonument(@PathVariable("id") Integer id,
     }
 
     /**
-     * Get all of the Monuments
+     * Get all of the Monuments and is active or inactive depending on onlyActive
+     * @param onlyActive - If true, only active monuments will be returned. If false, monuments
+     *                 will be returned regardless of whether they're active or inactive.
+     *                 If this is false then the user must be a partner or researcher to view
```
Comment (njd9419) [inline src/main/java/com/monumental/controllers/MonumentController.java]: Can you correct this to be `If this is false then the user must be a partner or above to view`?

File: src/main/java/com/monumental/controllers/MonumentController.java [hunk 17]
```diff
@@ -72,12 +98,22 @@ public Monument getMonument(@PathVariable("id") Integer id,
...
      * @return List<Monument> - List of all of the Monuments
+     * @throws AccessDeniedException - If trying to get inactive monuments without being a partner or researcher
```
Comment (njd9419) [inline src/main/java/com/monumental/controllers/MonumentController.java]: Same here: `If trying to get inactive monuments without being a partner or above`.

File: src/main/java/com/monumental/controllers/MonumentController.java [hunk 18]
```diff
@@ -92,6 +128,35 @@ public Monument updateMonument(@PathVariable("id") Integer id, @RequestBody Upda
         return this.monumentService.updateMonument(id, newMonument);
     }
 
+    private static class ToggleIsActiveRequest {
```
Comment (njd9419) [inline src/main/java/com/monumental/controllers/MonumentController.java]: Do you think it's worth it to add a comment how about why we decided to use a `private static class` vs. creating a new Request class for it?

File: src/main/java/com/monumental/controllers/MonumentController.java [hunk 19]
```diff
@@ -92,6 +128,35 @@ public Monument updateMonument(@PathVariable("id") Integer id, @RequestBody Upda
...
+
+    /**
+     * Change the value of isActive on a Monument
+     * @param id - ID of the Monument to update
+     * @param request - ToggleIsActiveRequest containing the new value for isActive
+     * @return Monument - The updated Monument
+     */
+    @PutMapping("/api/monument/active/{id}")
+    @PreAuthorize(Authorization.isResearcherOrAbove)
+    public Monument updateMonument(@PathVariable("id") Integer id, @RequestBody ToggleIsActiveRequest request) {
```
Comment (njd9419) [inline src/main/java/com/monumental/controllers/MonumentController.java]: Maybe we should rename this to `updateMonumentIsActive`?

Also, what's the different between `@RequestBody` and `@ModelAttribute`?

File: src/main/java/com/monumental/controllers/SearchController.java [hunk 20]
```diff
@@ -37,7 +44,10 @@
      * @param longitude - The longitude of the comparison point
      * @param distance - The distance from the comparison point to search in, units of miles
      * @param sortType - The way in which to sort the results by
+     * @param onlyActive - If true, only active monuments will be searched. If false, both inactive and active will be searched
      * @return            Matching Monuments based on their title, artist or location
+     * @throws AccessDeniedException - If trying to search for inactive monuments without being a partner or researcher
```
Comment (njd9419) [inline src/main/java/com/monumental/controllers/SearchController.java]: Can you correct this to be `If trying to search for inactive monuments without being a partner or above`?

File: src/main/java/com/monumental/services/MonumentService.java [hunk 21]
```diff
@@ -305,19 +311,20 @@ private void buildSearchQuery(CriteriaBuilder builder, CriteriaQuery query, Root
...
         CriteriaBuilder builder = this.getCriteriaBuilder();
         CriteriaQuery<Monument> query = this.createCriteriaQuery(builder, false);
         Root<Monument> root = this.createRoot(query);
         query.select(root);
 
         this.buildSearchQuery(
             builder, query, root, searchQuery, threshold, latitude, longitude, distance, tags, materials, sortType,
-            start, end, decade
+            start, end, decade, onlyActive
         );
```
Comment (njd9419) [inline src/main/java/com/monumental/services/MonumentService.java]: It makes me so happy our searching is extensible like this.

File: src/test/java/com/monumental/services/integrationtest/MonumentServiceMockIntegrationTests.java [hunk 22]
```diff
@@ -93,11 +93,11 @@ private MonumentBulkValidationResult validateCSV(String csvRows) {
...
         Mockito.doReturn(new ArrayList<>()).when(this.monumentServiceMock).findDuplicateMonuments(any(String.class),
-                any(Double.class), any(Double.class), any(String.class));
+                any(Double.class), any(Double.class), any(String.class), any(boolean.class));
         Mockito.doReturn(new ArrayList<>()).when(this.monumentServiceMock).findDuplicateMonuments(any(String.class),
-                isNull(), isNull(), any(String.class));
+                isNull(), isNull(), any(String.class), any(boolean.class));
         Mockito.doReturn(new ArrayList<>()).when(this.monumentServiceMock).findDuplicateMonuments(any(String.class),
-                any(Double.class), any(Double.class), isNull());
+                any(Double.class), any(Double.class), isNull(), any(boolean.class));
     }
```
Comment (njd9419) [inline src/test/java/com/monumental/services/integrationtest/MonumentServiceMockIntegrationTests.java]: Shouldn't these be `any(Boolean.class)` or does it not make a difference?

**Comment by njd9419** [comment]
First round of review done.

File: src/main/java/com/monumental/controllers/MonumentController.java (hunk 18 above)
Comment (benvogler) [reply src/main/java/com/monumental/controllers/MonumentController.java]: Not really. It's a simple wrapper. I think we'll be seeing more of these anyway

File: src/main/java/com/monumental/controllers/MonumentController.java (hunk 19 above)
Comment (benvogler) [reply src/main/java/com/monumental/controllers/MonumentController.java]: https://stackoverflow.com/a/31990142/10044594

Comment (njd9419) [reply src/main/java/com/monumental/controllers/MonumentController.java]: I thought `@RequestParam` was for query params

**Pull request merged**